ethcore/src/trace/db.rs

```diff
@@ -0,0 +1,573 @@
+	// cache
+	traces: RwLock<HashMap<H256, BlockTraces>>,
+	blooms: RwLock<HashMap<TraceGroupPosition, BlockTracesBloomGroup>>,
+	// db
+	tracesdb: Database,
+	// config,
+	config: Config,
+	// extras
+	extras: Arc<T>,
+}
+
+impl<T> BloomGroupDatabase for Tracedb<T> where T: DatabaseExtras {
+	fn blooms_at(&self, position: &GroupPosition) -> Option<BloomGroup> {
+		let position = TraceGroupPosition::from(position.clone());
+		self.tracesdb.read_with_cache(&self.blooms, &position).map(Into::into)
+	}
+}
+
+impl<T> Tracedb<T> where T: DatabaseExtras {
+	/// Creates new instance of `Tracedb`.
+	pub fn new(mut config: Config, path: &Path, extras: Arc<T>) -> Self {
+		let mut tracedb_path = path.to_path_buf();
+		tracedb_path.push("tracedb");
+		let tracesdb = Database::open_default(tracedb_path.to_str().unwrap()).unwrap();
+
+		// check if in previously tracing was enabled
+		let tracing_was_enabled = match tracesdb.get(b"enabled").unwrap() {
+			Some(ref value) if value as &[u8] == &[0x1] => Some(true),
+			Some(ref value) if value as &[u8] == &[0x0] => Some(false),
+			Some(_) => { panic!("tracesdb is malformed") },
+			None => None,
+		};
+
+		// compare it with the current option.
+		let tracing = match (tracing_was_enabled, config.enabled) {
+			(Some(true), Some(true)) => true,
+			(Some(true), None) => true,
+			(Some(true), Some(false)) => false,
+			(Some(false), Some(true)) => { panic!("Tracing can't be enabled. Resync required."); },
+			(Some(false), None) => false,
+			(Some(false), Some(false)) => false,
+			(None, Some(true)) => true,
+			_ => false,
+		};
+
+		config.enabled = Some(tracing);
+
+		let encoded_tracing= match tracing {
+			true => [0x1],
+			false => [0x0]
+		};
+
+		tracesdb.put(b"enabled", &encoded_tracing).unwrap();
+
+		Tracedb {
+			traces: RwLock::new(HashMap::new()),
+			blooms: RwLock::new(HashMap::new()),
+			tracesdb: tracesdb,
+			config: config,
+			extras: extras,
+		}
+	}
+
+	/// Returns traces for block with hash.
+	fn traces(&self, block_hash: &H256) -> Option<BlockTraces> {
+		self.tracesdb.read_with_cache(&self.traces, block_hash)
+	}
+
+	fn matching_block_traces(
+		&self,
+		filter: &Filter,
+		traces: FlatBlockTraces,
+		block_hash: H256,
+		block_number: BlockNumber
+	) -> Vec<LocalizedTrace> {
+		let tx_traces: Vec<FlatTransactionTraces> = traces.into();
+		tx_traces.into_iter()
+			.enumerate()
+			.flat_map(|(tx_number, tx_trace)| {
+				self.matching_transaction_traces(filter, tx_trace, block_hash.clone(), block_number, tx_number)
+			})
+			.collect()
+	}
+
+	fn matching_transaction_traces(
+		&self,
+		filter: &Filter,
+		traces: FlatTransactionTraces,
+		block_hash: H256,
+		block_number: BlockNumber,
+		tx_number: usize
+	) -> Vec<LocalizedTrace> {
+		let tx_hash = self.extras.transaction_hash(block_number, tx_number)
+		.expect("Expected to find transaction hash. Database is probably malformed");
+
+		let flat_traces: Vec<FlatTrace> = traces.into();
+		flat_traces.into_iter()
```

Comment:
I can't do that, cause flat trace does not have all required data - `trace_number`, `transaction_number`, `transaction_hash`, `block_number` and `block_hash`.